native/dlasy2.go

```diff
@@ -0,0 +1,270 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package native
+
+import (
+	"math"
+
+	"github.com/gonum/blas/blas64"
+)
+
+// Dlasy2 solves the Sylvester matrix equation where the matrices are of order 1
+// or 2. It computes the unknown n1×n2 matrix X so that
+//  TL*X   + sgn*X*TR   = scale*B,  if tranl == false and tranr == false,
+//  TL^T*X + sgn*X*TR   = scale*B,  if tranl == true  and tranr == false,
+//  TL*X   + sgn*X*TR^T = scale*B,  if tranl == false and tranr == true,
+//  TL^T*X + sgn*X*TR^T = scale*B,  if tranl == true  and tranr == true,
+// where TL is n1×n1, TR is n2×n2, B is n1×n2, and 1 <= n1,n2 <= 2.
+//
+// isgn must be 1 or -1, and n1 and n2 must be 0, 1, or 2, but these conditions
+// are not checked.
+//
+// Dlasy2 returns a scale factor that is chosen less than or equal to 1 to
+// prevent the solution overflowing, the infinity norm of the solution, and an
+// indicator of success. If ok is false, TL and TR have too close eigenvalues,
+// so TL or TR is perturbed to get a non-singular equation.
+//
+// Dlasy2 is an internal routine. It is exported for testing purposes.
+func (impl Implementation) Dlasy2(tranl, tranr bool, isgn, n1, n2 int, tl []float64, ldtl int, tr []float64, ldtr int, b []float64, ldb int, x []float64, ldx int) (scale, xnorm float64, ok bool) {
+	ok = true
+	// Quick return if possible.
+	if n1 == 0 || n2 == 0 {
+		return scale, xnorm, ok
+	}
+
+	// Set constants to control overflow.
+	eps := dlamchP
+	smlnum := dlamchS / eps
+	sgn := float64(isgn)
+
+	if n1 == 1 && n2 == 1 {
+		// 1×1 case: TL11*X + sgn*X*TR11 = B11.
+		tau1 := tl[0] + sgn*tr[0]
+		bet := math.Abs(tau1)
+		if bet <= smlnum {
+			tau1 = smlnum
+			bet = smlnum
+			ok = false
+		}
+		scale = 1
+		gam := math.Abs(b[0])
+		if smlnum*gam > bet {
+			scale = 1 / gam
+		}
+		x[0] = b[0] * scale / tau1
+		xnorm = math.Abs(x[0])
+		return scale, xnorm, ok
+	}
+
+	if n1+n2 == 3 {
+		// 1×2 or 2×1 case.
+		var (
+			smin float64
+			tmp  [4]float64 // tmp is used as a 2×2 column-major matrix.
```

Comment:
Would you look at the performance impact of making this a `[2][2]float64`? If it's not too great it would greatly clarify this var's use.